src/test/java/org/commcare/formplayer/auth/HmacAuthTests.java

```diff
@@ -120,9 +122,10 @@ public void testFullAuthEndpoint_WithHmacAuthAndUserDetails_Succeeds() throws Ex
 
     @Test
     public void testFullAuthMultipartEndpoint_WithHmacAuthAndUserDetails_Succeeds() throws Exception {
-        String hmac = RequestUtils.getHmac(formplayerAuthKey, FULL_AUTH_BODY);
-        MockHttpServletRequestBuilder builder = getMultipartRequestBuilder(getClass(), FULL_AUTH_BODY)
-                .header(Constants.HMAC_HEADER, hmac);
+        MockHttpServletRequestBuilder builder = getMultipartRequestBuilder(getClass(), FULL_AUTH_BODY);
+        MockHttpServletRequest request = builder.buildRequest(new MockServletContext());
+        String hmac = RequestUtils.getHmac(formplayerAuthKey, RequestUtils.getBody(request.getInputStream()));
```

Comment:
This looked a bit suspicious to me so I dug into it and found that `MockHttpServletRequest.getInputStream()` is returning an empty input stream.

I tried to figure out how we could do this better but there doesn't seem to be an easy way. I was able to hack it together by creating a real mutlpart body and adding that to the mock request as well as the individual parts:

```java
        MultipartBodyBuilder builder = new MultipartBodyBuilder();
        builder.part(PART_ANSWER, FULL_AUTH_BODY).contentType(MediaType.APPLICATION_JSON);

        Resource image = new ClassPathResource("media/valid_image.jpg");
        builder.part(PART_FILE, image).contentType(MediaType.IMAGE_JPEG);

        MultiValueMap<String, HttpEntity<?>> multipartBody = builder.build();
        RequestCallback callback = restTemplate.httpEntityCallback(multipartBody);
        MockClientHttpRequest request = new MockClientHttpRequest();
        callback.doWithRequest(request);

        String multiPartBody = request.getBodyAsString();

        MockPart answer = new MockPart(PART_ANSWER, FULL_AUTH_BODY.getBytes());
        answer.getHeaders().setContentType(MediaType.APPLICATION_JSON);
        MockHttpServletRequestBuilder request1 = multipart(String.format(String.format("/%s", "mock_mulipart_request")))
                .file(
                        new MockMultipartFile(
                                PART_FILE,
                                "valid_image.jpg",
                                MediaType.IMAGE_JPEG_VALUE,
                                image.getInputStream())
                )
                .part(answer)
                .contentType(request.getHeaders().get("Content-Type").get(0))
                .content(multiPartBody)
                .with(SecurityMockMvcRequestPostProcessors.csrf());

        String hmac = RequestUtils.getHmac(formplayerAuthKey, multiPartBody);
        request1.header(Constants.HMAC_HEADER, hmac);
        this.testEndpoint(request1, status().isOk());
```

I really don't like this approach though. I'm not sure what the best thing to do here. On the one hand we could disable multipart requests with HMAC auth (we don't currently have a use case for it) or we could abandon the idea of testing it properly for now.

Reply:
Ahh that's interesting! Thanks for digging deep into this. I would support disabling multipart requests with HMAC auth  if we don't see a plausible use case for it. I am assuming it would be comparatively low effort to do so and a cleaner approach amongst all.

Reply:
> Ahh that's interesting! Thanks for digging deep into this. I would support disabling multipart requests with HMAC auth if we don't see a plausible use case for it. I am assuming it would be comparatively low effort to do so and a cleaner approach amongst all.

👍🏻 HMAC auth should only be used for SMS forms which don't support MM questions

Reply:
Sounds promising. Do we need to make a HQ change as well for it ? Or Just changing FP to bypass Hmac auth on multipart requests should be enough ?